### ISSUE TITLE
Language settings SS/RA are not forwarded to Tiqr pages, which causes unwanted language switch

### ISSUE THREAD

**Issue by phavekes** [OpenConext/Stepup-tiqr#235]
[This issue is imported from pivotal - Originaly created at Feb 12, 2016 by Eefje van der Harst](https://www.pivotaltracker.com/story/show/113611545)

	In Self-Service:
- I start enrollment of 2nd factor in Self Service and language settings are: NL
- I select Tiqr 
- Screen \'Register new Tiqr account\' appears, but language is now set in EN 
https://tiqr.pilot.stepup.surfconext.nl/tiqr/enrol?return=https://tiqr.pilot.stepup.surfconext.nl/saml/sso_return

In RA Management:
- I start with entering activation code
- I click \'Verifieren bij Tiqr\'
- Screen \'Login with Tiqr\' appears, but language is now set in EN 
https://tiqr.pilot.stepup.surfconext.nl/tiqr/login?return=https://tiqr.pilot.stepup.surfconext.nl/saml/sso_return

Also: Inholland users complain that at first no QR code appears, so they cannot continue enrollment. 
Only after they switch language back to NL, a QR code appears and they can continue.
(zie bijlage)

Edit regarding locales (see dialogue below):
Gateway determines it locale solely based on the identity\'s SecondFactor. The displayed language is currently determined by the locale of SelfService and RA, which are stored in their sessions. MW/GW don\'t know about these locales upon entry. Furthermore, GW does not store any state about the identity. This is why the LocaleCookieListener is implemented in RA and in SelfService (initial contact) as well as in Gateway (subsequent contact).

**Comment by phavekes** [comment]
(Eefje van der Harst - Feb 12, 2016)

**Comment by phavekes** [comment]
@jodi or @pmeulen do you still want to use a cookie when SURFconext is switching to Accept-Language + cookie? (Boy Baukema - May 30, 2016)

**Comment by phavekes** [comment]
@boybaukema Keep in mind that in SSA the preferred locale of an identity is stored.  (Daan van Renterghem - May 30, 2016)

**Comment by phavekes** [comment]
SSA needs to know the preferred language of the user without the user being present to be able to send emails in the users preferred language.

The cookie should mirror the language preference that is stored for the user in the middleware. I don\'t think we should introduce another rule for the RA and the SS for determening the user\'s preferred language. The cookie is used only to communicate the language preference from the stepup gateway to the tiqr IdP or any Generic SAML Stepup Provider (GSSP).
 (Pieter van der Meulen - May 30, 2016)

**Comment by phavekes** [comment]
not working and even more language issues: screens with mix of NL and EN text (Ton Verschuren - Sep 20, 2016)

**Comment by phavekes** [comment]
@DRvR in SSA (Ton Verschuren - Sep 20, 2016)

**Comment by phavekes** [comment]
@tonverschuren could you list the screens (ideally with screenshots) where this is the case? (Daan van Renterghem - Sep 20, 2016)

**Comment by phavekes** [comment]
@DRvR please refer to attached screen dump; the previous page was in NL (Ton Verschuren - Sep 20, 2016)

**Comment by phavekes** [comment]
@tonverschuren This feature requires integration with tiqr, which afaik is not yet in place @jodi knows status (Pieter van der Meulen - Sep 23, 2016)

**Comment by phavekes** [comment]
@DRvR Is there documentation of this feature somewhere? What is the cookie used? Is ther configuration required? (Pieter van der Meulen - Sep 23, 2016)

**Comment by phavekes** [comment]
Indeed, this seems like a Tiqr issue.
The cookie used is configured in the [Stepup-bundle](https://github.com/SURFnet/Stepup-bundle/blob/4c2360dd757b36ef6c03278e61204c84745cb4cd/src/DependencyInjection/Configuration.php#L219-L253).  (Alex Rothuis - Oct 6, 2016)

**Comment by phavekes** [comment]
@alexrothuis It is correct that Tiqr does not listen to the cookie yet, something that @tonverschuren initially was not aware of. Implementing this in Stepup-tiqr is out of scope for this issue.

Based on the config in [Stepup-bundle](https://github.com/SURFnet/Stepup-bundle/blob/4c2360dd757b36ef6c03278e61204c84745cb4cd/src/DependencyInjection/Configuration.php#L219-L253) I expect a cookie with the name "stepup_locale" to be set after the user switches language.

Testing this with sa.test.surfconext.nl (parameters.yml: locale_cookie_domain: test.surfconext.nl) I do not see a "stepup_locale cookie" being set. This seems wrong. Is this a configuration issue, a bug in the code or am I misunderstanding the functionality? (Pieter van der Meulen - Oct 7, 2016)

**Comment by phavekes** [comment]
I have made PRs against SelfService and RA to implement the LocaleCookieListener from the Stepup-bundle. This should make the locale available (externally) using the cookie.  (Alex Rothuis - Oct 11, 2016)

**Comment by phavekes** [comment]
Ah, I see it now. Although the PRs against SelfService and RA can be used to set an external cookie when the locales are changes there, the original approach was to set the cookie during a visit to Gateway. 

We have to make sure the cookie settings are correct there, before merging unnecessary PRs. Do you have the `locale_cookie_domain` set correctly in Gateway? (Alex Rothuis - Oct 12, 2016)

**Comment by phavekes** [comment]
Indeed the reasoning was that because all requests to Tiqr use SAML authentication via the gateway it would be sufficient to update the cookie there. Same goes for any other "external" 2nd factors, which would have the same issue as tiqr in finding the preferred language of the user. Since these all would use GSSP as well, they would always pass the gateway. The only requirement would be that the services are hosted within the "locale_cookie_domain". (Pieter van der Meulen - Oct 12, 2016)

**Comment by phavekes** [comment]
I think I have figured out why the cookie is not being set by Gateway: a preferred locale cannot yet be determined when a second factor is being selected in Gateway, as at that point an [intrinsic Loa](https://github.com/SURFnet/Stepup-Gateway/blob/8f3ee746814759319bcb64ebc3814982bb2884be/src/Surfnet/StepupGateway/GatewayBundle/Controller/SecondFactorController.php#L70-L74) is used, while [setting the preferred locale](https://github.com/SURFnet/Stepup-Gateway/blob/8f3ee746814759319bcb64ebc3814982bb2884be/src/Surfnet/StepupGateway/GatewayBundle/Controller/SecondFactorController.php#L109) is determined based on a (known) [viable second factor](https://github.com/SURFnet/Stepup-Gateway/blob/8f3ee746814759319bcb64ebc3814982bb2884be/src/Surfnet/StepupGateway/GatewayBundle/Controller/SecondFactorController.php#L76-L78). So Gateway determines it locale solely based on the identity\'s SecondFactor. It does not talk to MW by design.

The displayed language is currently determined by the locale of SelfService and RA, which are stored in their sessions. MW/GW don\'t know about these locales upon entry. Furthermore, GW does not store any state about the identity. If we implement the locale cookie listener in [RA](https://github.com/SURFnet/Stepup-RA/pull/123) and in [SelfService](https://github.com/SURFnet/Stepup-SelfService/pull/111)  the locale cookie is set separate from but comparable to how they\'re shown in RA and SelfService. 

However, this means if the request is not coming from RA or SelfService, we do not have a guarantee the cookie is set. This means that if we ever were to create another \'front-end\', we have to implement the listener there as well. The StepupBundle already exposes an interface and a service to make this easy. (Alex Rothuis - Oct 13, 2016)

**Comment by phavekes** [comment]
As discussed:
- Implemented/added a locale listener in SS and RA. 
- Keep the locale listener in the GW.

The locale listener sets/updates the "stepup_locale" (configurable) cookie of the user for a domain (configurable). This allows an IdP using a GSSP proxy (I.e. tiqr) to use the "stepup_locale" cookie to get the preferred language of the user. The IdP must 

During registration:  (Pieter van der Meulen - Oct 17, 2016)
